File: CrackingTheCodingInterview/ch04-TreesAndGraphs/minimalTree.js
/*
  given a sorted array with unique integer elements. write an
  algorithm to create a BST with minimal height
*/

function minimalBST(arr) {
  if (arr.length === 0) return null;
  if (arr.length === 1) return { value: arr[0] };
  // set root to midpoint of array
  const middle = Math.floor((arr.length - 1) / 2);

  // each node of BST is an object with
  // value, left and right properties

  // recursively build tree using each half as left and right
  // child BST's
  return {
    value: arr[middle],
    left: minimalBST(arr.slice(0, middle)),
    right: minimalBST(arr.slice(middle + 1))
  }
}

const bst = minimalBST([1,2,3,5,7,9]);
/*
      3
    /   \
  1      7
  \     / \
   2   5   9
*/

console.log(bst.value) //=> 3
console.log(bst.left.value) //=> 1
console.log(bst.right.value) //=> 7
console.log(bst.left.right.value) //=> 2
console.log(bst.right.left.value) //=> 5
console.log(bst.right.right.value) //=> 9

File: CrackingTheCodingInterview/ch04-TreesAndGraphs/minimalTree.py
# Given a sorted array with unique integer elements. Write an
# algorithm to create a BST with minimal height
import math


class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def assign_left(self, node):
        self.left = node

    def assign_right(self, node):
        self.right = node


def minimal_tree(arr):
    if len(arr) == 0:
        return None
    if len(arr) == 1:
        return Node(arr[0])

    # Find middle
    m = math.floor(len(arr) / 2)

    # Recursively build left and right
    root = Node(arr[m])
    root.assign_left(minimal_tree(arr[0:m]))
    root.assign_right(minimal_tree(arr[m+1:]))

    return root


